cmd/libsnap-confine-private/device-cgroup-support.c

```diff
@@ -107,6 +111,322 @@ static void _sc_cgroup_v1_attach_pid(sc_device_cgroup *self, pid_t pid) {
+        BPF_JMP_A(2),
+        /* unknown device type */
+        BPF_MOV64_IMM(BPF_REG_0, 0), /* r0 = 0 */
+        BPF_EXIT_INSN(),
+        /* back on happy path, prepare arguments for map lookup */
+        BPF_LD_MAP_FD(BPF_REG_1, map_fd),
+        BPF_MOV64_REG(BPF_REG_2, BPF_REG_6),                                 /* r2 = (struct key *) r6, */
+        BPF_RAW_INSN(BPF_JMP | BPF_CALL, 0, 0, 0, BPF_FUNC_map_lookup_elem), /* r0 = bpf_map_lookup_elem(<map>,
+                                                                                &key) */
+        BPF_JMP_IMM(BPF_JEQ, BPF_REG_0, 0, 1),                               /* if (value_ptr == 0) goto pc + 1 */
+        /* we found an exact match */
+        BPF_JMP_A(5), /* else goto pc + 5 */
+        /* maybe the minor number is using 0xffffffff (any) mask */
+        BPF_ST_MEM(BPF_W, BPF_REG_6, offsetof(struct sc_cgroup_v2_device_key, minor), UINT32_MAX),
+        BPF_LD_MAP_FD(BPF_REG_1, map_fd),
+        BPF_MOV64_REG(BPF_REG_2, BPF_REG_6),                                 /* r2 = (struct key *) r6, */
+        BPF_RAW_INSN(BPF_JMP | BPF_CALL, 0, 0, 0, BPF_FUNC_map_lookup_elem), /* r0 = bpf_map_lookup_elem(<map>,
+                                                                                &key) */
+        BPF_JMP_IMM(BPF_JEQ, BPF_REG_0, 0, 2),                               /* if (value_ptr == 0) goto pc + 2 */
+        /* we found a match with any minor number for that type|major */
+        BPF_MOV64_IMM(BPF_REG_0, 1), /* r0 = 1 */
+        BPF_JMP_A(1),
+        BPF_MOV64_IMM(BPF_REG_0, 0), /* r0 = 0 */
+        BPF_EXIT_INSN(),
+    };
+
+    char log_buf[4096] = {0};
+
+    int prog_fd =
+        bpf_load_prog(BPF_PROG_TYPE_CGROUP_DEVICE, prog, sizeof(prog) / sizeof(prog[0]), log_buf, sizeof(log_buf));
+    if (prog_fd < 0) {
+        die("cannot load program:\n%s\n", log_buf);
+    }
+    return prog_fd;
+}
+
+static int _sc_cgroup_v2_init_bpf(sc_device_cgroup *self, int flags) {
+    self->v2.devmap_fd = -1;
+    self->v2.cgroup_fd = -1;
+
+    char *own_group SC_CLEANUP(sc_cleanup_string) = sc_cgroup_v2_own_path_full();
+    if (own_group == NULL) {
+        die("cannot obtain own group path");
+    }
+
+    char own_group_full[PATH_MAX] = {0};
+    sc_must_snprintf(own_group_full, sizeof(own_group_full), "/sys/fs/cgroup/%s", own_group);
+    int cgroup_fd = open(own_group_full, O_PATH | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+    if (cgroup_fd < 0) {
+        die("cannot open own cgroup directory %s", own_group_full);
+    }
+    debug("cgroup %s opened at %d", own_group_full, cgroup_fd);
+
+    self->v2.tag = sc_strdup(self->security_tag);
+    /* bpffs is unhappy about dots in the name, replace all with underscores */
+    for (char *c = strchr(self->v2.tag, '.'); c != NULL; c = strchr(c, '.')) {
+        *c = '_';
+    }
+
+    char path[PATH_MAX] = {0};
+    sc_must_snprintf(path, sizeof path, "/sys/fs/bpf/snap/%s", self->v2.tag);
```

Comment:
somewhere, maybe here could you explain the concept of pinning a BPF map object to this specific path and how that's important?